tests/integration/test_charm_no_runner.py

```diff
@@ -191,3 +192,29 @@ async def test_reconcile_runners(model: Model, app_no_runner: Application) -> No
     await reconcile(app=app, model=model)
 
     await wait_till_num_of_runners(unit, 0)
+
+
+@pytest.mark.asyncio
+@pytest.mark.abort_on_fail
+async def test_charm_upgrade(model: Model, app_no_runner: Application, charm_file: str) -> None:
+    """
+    arrange: A working application with no runners.
+    act: Upgrade the charm.
+    assert: The upgrade_charm hook ran successfully.
+    """
+    await app_no_runner.refresh(path=charm_file)
+
+    # It may take some time between the charm upgrade and the run of the upgrade charm hook,
+    # so set the idle_period high.
+    await model.wait_for_idle(status=ACTIVE, idle_period=60)
+    unit = app_no_runner.units[0]
+    unit_name_without_slash = unit.name.replace("/", "-")
+    ret_code, stdout = await run_in_unit(
+        unit=unit, command=f"cat /var/log/juju/unit-{unit_name_without_slash}.log"
+    )
+    assert ret_code == 0, f"Failed to read the log file: {stdout}"
+    assert stdout, "Log file is empty."
+    assert "Emitting Juju event upgrade_charm." in stdout, "upgrade_charm event not yet fired."
+    # There may be a race condition: if the hook is executed shortly after idle,
+    # it may not have finished yet, so check for idle again.
```

Comment:
I'm wondering if the `helpers.wait_for` function might help here for the log message to appear before doing the assertions.